File: server/controller/downloader.ts
import { Request, Response } from "express"
import { exec } from "child_process"

interface VideoMetaData {
	title: string;
	thumbnail: string;  
	duration: number;      
	category?: string;    
	links: {link: string, quality: string}[];
}

const extractMetaData = (url: string): Promise<VideoMetaData> => {
  return new Promise((resolve, reject) => {
    exec(`yt-dlp -j --no-playlist ${url}`, (error, stdout) => {
      if(error) return reject(`Failed to fetch meta data ${error.message}`)
      
      try{
        const metaData = JSON.parse(stdout)
        const videoData: VideoMetaData = {
          title: metaData.title || "No Title",
          thumbnail: metaData.thumbnail || "",
          duration: metaData.duration || 0,
          category: metaData.categories?.[0] || "Uncharacterized",
          links: metaData.formats.filter((format: any) => format.url).map((format: any) => ({
            link: format.url,
            quality: format.format,
          })),
        }
        resolve(videoData)
      } catch(error){
        reject(`Error parsing Data ${error}`)
      }
    })
  })
}

export const fetchMetaData = async (req: Request, res: Response): Promise<any> => {
  const videourl = req.query.url as string 
  if(!videourl) return res.status(400).json({error: "Video Url is Required"})
  
  try{
    const metaData = await extractMetaData(videourl)
    res.status(200).json(metaData)
  } catch(error){
    res.status(500).json({error: "Failed to get metadata"})
    console.error(error)
  }
}

File: server/controller/downloader.go
package controller

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os/exec"
	"github.com/gin-gonic/gin"
)

type VideoMetaData struct {
	Title string `json:"title"`
	Thumbnail string `json:"thumbnail"`
	Duration int `json:"duration"`
	Category string `json:"category"`
	Links []VideoQuality `json:"links"`
}


type VideoQuality struct {
	Link    string `json:"link"`
	Quality string `json:"quality"`
}

func ExtractMetaData(videoURL string) (*VideoMetaData, error) {
	cmd := exec.Command("yt-dlp", "-j", "--no-playlist", videoURL)
	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("failed to fetch metadata: %v", err)
	}

	var metaData map[string]interface{}
	if err := json.Unmarshal(output, &metaData); err != nil {
		return nil, fmt.Errorf("error parsing data: %v", err)
	}

	var links []VideoQuality
	formats, ok := metaData["formats"].([]interface{})
	if ok {
		for _, format := range formats {
			f := format.(map[string]interface{})
			if f["url"] != nil {
				links = append(links, VideoQuality{
					Link:    f["url"].(string),
					Quality: f["format"].(string),
				})
			}
		}
	}

	videoData := &VideoMetaData{
		Title:     metaData["title"].(string),
		Thumbnail: metaData["thumbnail"].(string),
		Duration:  int(metaData["duration"].(float64)),
		Category:  "Uncategorized",
		Links:     links,
	}

	if categories, ok := metaData["categories"].([]interface{}); ok && len(categories) > 0 {
		videoData.Category = categories[0].(string)
	}

	return videoData, nil
}

func ProxyRequest(c *gin.Context) {
	targetURL := c.Query("url")
	if targetURL == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Missing URL parameter"})
		return
	}

	parsedURL, err := url.Parse(targetURL)
	if err != nil || !(parsedURL.Scheme == "http" || parsedURL.Scheme == "https") {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid URL"})
		return
	}

	resp, err := http.Get(targetURL)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch resource"})
		return
	}
	defer resp.Body.Close()

	c.Header("Content-Type", resp.Header.Get("Content-Type"))
	c.Header("Cache-Control", "public, max-age=86400")
	io.Copy(c.Writer, resp.Body)
}